### PR TITLE
 Cloud Connection Status, Cloud app retry sequence, and Cloud app activation flow

### PR DESCRIPTION
**Pull request by jacobkeeler**
Partially Implements #2215 

This PR is **not ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- TBD

### Summary
Adds RPC changes to the cloud application registration/activation flow, as well as implements the connection retry sequence for cloud apps.

##### Enhancements
* Adds `SDL.ActivateApp` flow changes for cloud apps
* Implements cloud app connection retry sequence
* Adds cloud connection status parameters to HMI application struct

### Tasks Remaining:
- [x] Add Cloud Connection Status implementation
- [x] Test Cloud Connection Status
- [x] Successfully build with `BUILD_TESTS=ON`
- [x] Add method headers and comments

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/sdl_activate_app_request.cc [hunk 1]
```diff
@@ -139,7 +145,7 @@ void SDLActivateAppRequest::Run() {
   const uint32_t application_id = app_id();
 
   ApplicationConstSharedPtr app_to_activate =
-      application_manager_.application(application_id);
+      application_manager_.application_by_hmi_app(application_id);
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/sdl_activate_app_request.cc]: For regular apps, the application_id is the internal app id, not the hmi app id. Core can't find the application when activating a normal app.

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc [hunk 2]
```diff
@@ -352,6 +363,7 @@ TransportAdapter::Error TransportAdapterImpl::DisconnectDevice(
   Error error = OK;
   DeviceSptr device = FindDevice(device_id);
   device->set_connection_status(ConnectionStatus::CLOSING);
+  ConnectionStatusUpdated();
```
Comment (jacobkeeler) [inline src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: Might be able to integrate this into the `set_connection_status` function, need to look into.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 3]
```diff
@@ -807,9 +833,56 @@ void ApplicationManagerImpl::CollectCloudAppInformation() {
 
     pending_device_map_.insert(
         std::pair<std::string, std::string>(endpoint, *it));
+    auto old_device = old_device_map.find(endpoint);
+    if (old_device_map.find(endpoint) != old_device_map.end()) {
```
Comment (JackLivio) [inline src/components/application_manager/src/application_manager_impl.cc]: Redundant find(). Change to `old_device != old_device_map.end()`

File: src/components/application_manager/src/application_manager_impl.cc [hunk 4]
```diff
@@ -807,9 +833,56 @@ void ApplicationManagerImpl::CollectCloudAppInformation() {
...
+      PolicyAppIdPredicate finder(policy_app_id);
+      ApplicationSet::iterator it = std::find_if(
+          apps_to_register_.begin(), apps_to_register_.end(), finder);
+      if (it == apps_to_register_.end()) {
+        continue;
+      }
+      app = *it;
+      apps_to_register_.erase(it);
+    }
+    if (app.use_count() == 0) {
```
Comment (JackLivio) [inline src/components/application_manager/src/application_manager_impl.cc]: This condition should be removed.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 5]
```diff
@@ -782,20 +803,25 @@ void ApplicationManagerImpl::OnHMIStartedCooperation() {
...
-  CollectCloudAppInformation();
+  RefreshCloudAppInformation();
 }
 
-void ApplicationManagerImpl::CollectCloudAppInformation() {
-  std::vector<std::string> cloud_app_id_vector;
-  GetPolicyHandler().GetEnabledCloudApps(cloud_app_id_vector);
-  std::vector<std::string>::iterator it = cloud_app_id_vector.begin();
-  std::vector<std::string>::iterator end = cloud_app_id_vector.end();
+void ApplicationManagerImpl::RefreshCloudAppInformation() {
```
Comment (JackLivio) [inline src/components/application_manager/src/application_manager_impl.cc]: This functions behavior and purpose is hard to understand at first. Function could use some more comments describing the flow.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 6]
```diff
@@ -782,20 +803,25 @@ void ApplicationManagerImpl::OnHMIStartedCooperation() {
...
+  LOG4CXX_AUTO_TRACE(logger_);
+  std::vector<std::string> enabled_apps;
+  GetPolicyHandler().GetEnabledCloudApps(enabled_apps);
+  std::vector<std::string>::iterator it = enabled_apps.begin();
```
Comment (JackLivio) [inline src/components/application_manager/src/application_manager_impl.cc]: Maybe rename to "enabled_it"

File: src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc (hunk 2 above)
Comment (jacobkeeler) [reply src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc]: Resolved with `UpdateConnectionStatus`

File: src/components/application_manager/src/application_manager_impl.cc (hunk 5 above)
Comment (jacobkeeler) [reply src/components/application_manager/src/application_manager_impl.cc]: Added comments

File: src/components/transport_manager/test/transport_adapter_test.cc [hunk 7]
```diff
@@ -439,18 +502,23 @@ TEST_F(TransportAdapterTest, DisconnectDevice_DeviceAddedConnectionCreated) {
   EXPECT_CALL(*serverMock, IsInitialised()).WillOnce(Return(true));
   EXPECT_CALL(*serverMock, CreateConnection(uniq_id, app_handle))
       .WillOnce(Return(TransportAdapter::OK));
-  EXPECT_CALL(transport_adapter, FindDevice(uniq_id)).WillOnce(Return(mockdev));
+  EXPECT_CALL(transport_adapter, FindDevice(uniq_id))
+      .WillRepeatedly(Return(mockdev));
+  // EXPECT_CALL(transport_adapter, ConnectionStatusUpdated(_, _));
```
Comment (jacobkeeler) [inline src/components/transport_manager/test/transport_adapter_test.cc]: Should be removed or re-added, not commented

File: src/components/transport_manager/test/transport_adapter_test.cc [hunk 8]
```diff
@@ -439,18 +502,23 @@ TEST_F(TransportAdapterTest, DisconnectDevice_DeviceAddedConnectionCreated) {
...
   EXPECT_EQ(TransportAdapter::OK, res);
+  EXPECT_EQ(ConnectionStatus::CONNECTED, mockdev->connection_status());
 
   auto mock_connection = std::make_shared<MockConnection>();
   transport_adapter.ConnectionCreated(mock_connection, uniq_id, app_handle);
 
   EXPECT_CALL(*mock_connection, Disconnect())
       .WillOnce(Return(TransportAdapter::OK));
 
+  // EXPECT_CALL(transport_adapter, ConnectionStatusUpdated(_, _));
```
Comment (jacobkeeler) [inline src/components/transport_manager/test/transport_adapter_test.cc]: Same here, remove or uncomment

File: src/components/transport_manager/test/include/transport_manager/transport_adapter/mock_transport_adapter_controller.h [hunk 9]
```diff
@@ -44,7 +44,7 @@ using namespace ::transport_manager::transport_adapter;
 
 class MockTransportAdapterController : public TransportAdapterController {
  public:
-  MOCK_METHOD1(AddDevice, DeviceSptr(DeviceSptr device));
+  MOCK_METHOD1(AddDevice, DeviceSptr(std::string device));
```
Comment (jacobkeeler) [inline src/components/transport_manager/test/include/transport_manager/transport_adapter/mock_transport_adapter_controller.h]: This change probably shouldn't compile, this does not match the method signature in the TransportAdapterController. It's possible that it just isn't used anywhere, but this change should be reverted anyway.

**Pull request merged**